plugins/nf-tower/src/main/io/seqera/tower/plugin/TowerFusionEnv.groovy

```diff
@@ -0,0 +1,257 @@
+            .version(HttpClient.Version.HTTP_1_1)
+            .followRedirects(HttpClient.Redirect.NEVER)
+            .cookieHandler(new CookieManager())
+            .connectTimeout(DEFAULT_CONNECTION_TIMEOUT)
+        // use virtual threads executor if enabled
+        if ( Threads.useVirtual() ) {
+            builder.executor(Executors.newVirtualThreadPerTaskExecutor())
+        }
+        // build and return the new client
+        return builder.build()
+    }
+
+    /**
+     * Create a new RetryPolicy instance with default settings and the given list of retryable errors. With this policy,
+     * a request is retried on IOExceptions and any server errors defined in errorsToRetry. The number of retries, delay,
+     * max delay, and jitter are controlled by the corresponding values defined at class level.
+     *
+     * @return The new RetryPolicy instance
+     */
+    private static <T> RetryPolicy<HttpResponse<T>> newDefaultRetryPolicy(List<Integer> errorsToRetry) {
+
+        final retryOnException = (e -> e instanceof IOException) as Predicate<? extends Throwable>
+        final retryOnStatusCode = ((HttpResponse<T> resp) -> resp.statusCode() in errorsToRetry) as Predicate<HttpResponse<T>>
+
+        final listener = new EventListener<ExecutionAttemptedEvent<HttpResponse<T>>>() {
+            @Override
+            void accept(ExecutionAttemptedEvent event) throws Throwable {
+                def msg = "connection failure - attempt: ${event.attemptCount}"
+                if (event.lastResult != null)
+                    msg += "; response: ${event.lastResult}"
+                if (event.lastFailure != null)
+                    msg += "; exception: [${event.lastFailure.class.name}] ${event.lastFailure.message}"
+                log.debug(msg)
+            }
+        }
+        return RetryPolicy.<HttpResponse<T>> builder()
+            .handleIf(retryOnException)
+            .handleResultIf(retryOnStatusCode)
+            .withBackoff(DEFAULT_RETRY_POLICY_DELAY.toMillis(), DEFAULT_RETRY_POLICY_MAX_DELAY.toMillis(), ChronoUnit.MILLIS)
+            .withMaxAttempts(DEFAULT_RETRY_POLICY_MAX_ATTEMPTS)
+            .withJitter(DEFAULT_RETRY_POLICY_JITTER)
+            .onRetry(listener)
+            .build()
+    }
+
+    /**
+     * Send an HTTP request and return the response. This method automatically retries the request according to the
+     * given RetryPolicy.
+     *
+     * @param req The HttpRequest to send
+     * @return The HttpResponse received
+     */
+    private <T> HttpResponse<String> safeHttpSend(HttpRequest req, RetryPolicy<T> policy) {
+        return Failsafe.with(policy).get(
+            () -> {
+                log.debug "Request: method:=${req.method()}; uri:=${req.uri()}; request:=${req}"
+                final resp = httpClient.send(req, HttpResponse.BodyHandlers.ofString())
+                log.debug "Response: statusCode:=${resp.statusCode()}; body:=${resp.body()}"
```

Comment:
This response is going to contain an authentication token right? Do we want to log it?

Reply:
Good question. I'm not sure about what the preferred policy is in Nextflow logs (cc @pditommaso)